File: app/src/main/java/com/owncloud/android/ui/dialog/RenameFileDialogFragment.java
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2022 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2017-2022 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2014 ownCloud Inc.
 * SPDX-FileCopyrightText: 2014 David A. Velasco <dvelasco@solidgear.es>
 * SPDX-License-Identifier: GPL-2.0-only AND (AGPL-3.0-or-later OR GPL-2.0-only)
 */
package com.owncloud.android.ui.dialog;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.common.collect.Sets;
import com.nextcloud.client.di.Injectable;
import com.nextcloud.utils.extensions.BundleExtensionsKt;
import com.owncloud.android.R;
import com.owncloud.android.databinding.EditBoxDialogBinding;
import com.owncloud.android.datamodel.FileDataStorageManager;
import com.owncloud.android.datamodel.OCFile;
import com.owncloud.android.lib.resources.files.FileUtils;
import com.owncloud.android.ui.activity.ComponentsGetter;
import com.owncloud.android.utils.DisplayUtils;
import com.owncloud.android.utils.KeyboardUtils;
import com.owncloud.android.utils.theme.ViewThemeUtils;

import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

/**
 *  Dialog to input a new name for an {@link OCFile} being renamed.
 *  Triggers the rename operation.
 */
public class RenameFileDialogFragment
    extends DialogFragment implements DialogInterface.OnClickListener, TextWatcher, Injectable {

    private static final String ARG_TARGET_FILE = "TARGET_FILE";
    private static final String ARG_PARENT_FOLDER = "PARENT_FOLDER";

    @Inject ViewThemeUtils viewThemeUtils;
    @Inject FileDataStorageManager fileDataStorageManager;
    @Inject KeyboardUtils keyboardUtils;

    private EditBoxDialogBinding binding;
    private OCFile mTargetFile;
    private MaterialButton positiveButton;
    private Set<String> fileNames;

    /**
     * Public factory method to create new RenameFileDialogFragment instances.
     *
     * @param file File to rename.
     * @return Dialog ready to show.
     */
    public static RenameFileDialogFragment newInstance(OCFile file, OCFile parentFolder) {
        RenameFileDialogFragment frag = new RenameFileDialogFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_TARGET_FILE, file);
        args.putParcelable(ARG_PARENT_FOLDER, parentFolder);
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onStart() {
        super.onStart();
        initAlertDialog();
    }

    @Override
    public void onResume() {
        super.onResume();
        keyboardUtils.showKeyboardForEditText(requireDialog().getWindow(), binding.userInput);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        mTargetFile = BundleExtensionsKt.getParcelableArgument(requireArguments(), ARG_TARGET_FILE, OCFile.class);

        // Inflate the layout for the dialog
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        binding = EditBoxDialogBinding.inflate(inflater, null, false);
        View view = binding.getRoot();

        // Setup layout
        String currentName = mTargetFile.getFileName();
        binding.userInput.setText(currentName);
        viewThemeUtils.material.colorTextInputLayout(binding.userInputContainer);
        int extensionStart = mTargetFile.isFolder() ? -1 : currentName.lastIndexOf('.');
        int selectionEnd = extensionStart >= 0 ? extensionStart : currentName.length();
        binding.userInput.setSelection(0, selectionEnd);

        OCFile parentFolder = BundleExtensionsKt.getParcelableArgument(getArguments(), ARG_PARENT_FOLDER, OCFile.class);
        List<OCFile> folderContent = fileDataStorageManager.getFolderContent(parentFolder, false);
        fileNames = Sets.newHashSetWithExpectedSize(folderContent.size());

        for (OCFile file : folderContent) {
            fileNames.add(file.getFileName());
        }

        // Add TextChangedListener to handle showing/hiding the input warning message
        binding.userInput.addTextChangedListener(this);

        // Build the dialog
        MaterialAlertDialogBuilder builder = buildMaterialAlertDialog(view);

        viewThemeUtils.dialog.colorMaterialAlertDialogBackground(binding.userInputContainer.getContext(), builder);

        return builder.create();
    }

    private MaterialAlertDialogBuilder buildMaterialAlertDialog(View view) {
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(requireActivity());

        builder
            .setView(view)
            .setPositiveButton(R.string.file_rename, this)
            .setNegativeButton(R.string.common_cancel, this)
            .setTitle(R.string.rename_dialog_title);

        return builder;
    }

    private void initAlertDialog() {
        AlertDialog alertDialog = (AlertDialog) getDialog();

        if (alertDialog != null) {
            positiveButton = (MaterialButton) alertDialog.getButton(AlertDialog.BUTTON_POSITIVE);
            MaterialButton negativeButton = (MaterialButton) alertDialog.getButton(AlertDialog.BUTTON_NEGATIVE);

            viewThemeUtils.material.colorMaterialButtonPrimaryTonal(positiveButton);
            viewThemeUtils.material.colorMaterialButtonPrimaryBorderless(negativeButton);
        }
    }

    @Override
    public void onClick(DialogInterface dialog, int which) {
        if (which == AlertDialog.BUTTON_POSITIVE) {
            String newFileName = "";

            if (binding.userInput.getText() != null) {
                newFileName = binding.userInput.getText().toString().trim();
            }

            if (TextUtils.isEmpty(newFileName)) {
                DisplayUtils.showSnackMessage(requireActivity(), R.string.filename_empty);
                return;
            }

            if (!FileUtils.isValidName(newFileName)) {
                DisplayUtils.showSnackMessage(requireActivity(), R.string.filename_forbidden_charaters_from_server);

                return;
            }

            ((ComponentsGetter) requireActivity()).getFileOperationsHelper().renameFile(mTargetFile, newFileName);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    /**
     * When user enters a hidden file name, the 'hidden file' message is shown.
     * Otherwise, the message is ensured to be hidden.
     */
    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        String newFileName = "";
        if (binding.userInput.getText() != null) {
            newFileName = binding.userInput.getText().toString().trim();
        }

        if (!TextUtils.isEmpty(newFileName) && newFileName.charAt(0) == '.') {
            binding.userInputContainer.setError(getText(R.string.hidden_file_name_warning));
        } else if (TextUtils.isEmpty(newFileName)) {
            binding.userInputContainer.setError(getString(R.string.filename_empty));
            positiveButton.setEnabled(false);
        } else if (fileNames.contains(newFileName)) {
            binding.userInputContainer.setError(getText(R.string.file_already_exists));
            positiveButton.setEnabled(false);
        } else if (binding.userInputContainer.getError() != null) {
            binding.userInputContainer.setError(null);
            // Called to remove extra padding
            binding.userInputContainer.setErrorEnabled(false);
            positiveButton.setEnabled(true);
        }
    }

    @Override
    public void afterTextChanged(Editable s) {

    }
}

File: app/src/main/java/com/owncloud/android/ui/dialog/RenameFileDialogFragment.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2022 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2017-2022 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2014 ownCloud Inc.
 * SPDX-FileCopyrightText: 2014 David A. Velasco <dvelasco@solidgear.es>
 * SPDX-License-Identifier: GPL-2.0-only AND (AGPL-3.0-or-later OR GPL-2.0-only)
 */
package com.owncloud.android.ui.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.google.android.material.button.MaterialButton
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.common.collect.Sets
import com.nextcloud.client.account.CurrentAccountProvider
import com.nextcloud.client.di.Injectable
import com.nextcloud.utils.extensions.getParcelableArgument
import com.nextcloud.utils.fileNameValidator.FileNameValidator.checkFileName
import com.nextcloud.utils.fileNameValidator.FileNameValidator.isFileHidden
import com.owncloud.android.R
import com.owncloud.android.databinding.EditBoxDialogBinding
import com.owncloud.android.datamodel.FileDataStorageManager
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.lib.resources.status.OCCapability
import com.owncloud.android.ui.activity.ComponentsGetter
import com.owncloud.android.ui.activity.FileDisplayActivity
import com.owncloud.android.utils.DisplayUtils
import com.owncloud.android.utils.KeyboardUtils
import com.owncloud.android.utils.theme.ViewThemeUtils
import javax.inject.Inject

/**
 * Dialog to input a new name for an [OCFile] being renamed.
 * Triggers the rename operation.
 */
class RenameFileDialogFragment : DialogFragment(), DialogInterface.OnClickListener, TextWatcher, Injectable {
    @Inject
    lateinit var viewThemeUtils: ViewThemeUtils

    @Inject
    lateinit var fileDataStorageManager: FileDataStorageManager

    @Inject
    lateinit var keyboardUtils: KeyboardUtils

    @Inject
    lateinit var currentAccount: CurrentAccountProvider

    private lateinit var binding: EditBoxDialogBinding
    private var mTargetFile: OCFile? = null
    private var positiveButton: MaterialButton? = null
    private var fileNames: MutableSet<String>? = null

    override fun onStart() {
        super.onStart()
        initAlertDialog()
    }

    override fun onResume() {
        super.onResume()
        keyboardUtils.showKeyboardForEditText(requireDialog().window, binding.userInput)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        mTargetFile = requireArguments().getParcelableArgument(ARG_TARGET_FILE, OCFile::class.java)

        val inflater = requireActivity().layoutInflater
        binding = EditBoxDialogBinding.inflate(inflater, null, false)

        val currentName = mTargetFile?.fileName
        binding.userInput.setText(currentName)
        viewThemeUtils.material.colorTextInputLayout(binding.userInputContainer)
        val extensionStart = if (mTargetFile?.isFolder == true) -1 else currentName?.lastIndexOf('.')
        val selectionEnd = if ((extensionStart ?: -1) >= 0) extensionStart else currentName?.length
        if (selectionEnd != null) {
            binding.userInput.setSelection(0, selectionEnd)
        }

        val parentFolder = arguments.getParcelableArgument(ARG_PARENT_FOLDER, OCFile::class.java)
        val folderContent = fileDataStorageManager.getFolderContent(parentFolder, false)
        fileNames = Sets.newHashSetWithExpectedSize(folderContent.size)

        for (file in folderContent) {
            fileNames?.add(file.fileName)
        }

        binding.userInput.addTextChangedListener(this)

        val builder = buildMaterialAlertDialog(binding.root)

        viewThemeUtils.dialog.colorMaterialAlertDialogBackground(binding.userInputContainer.context, builder)

        return builder.create()
    }

    private fun buildMaterialAlertDialog(view: View): MaterialAlertDialogBuilder {
        val builder = MaterialAlertDialogBuilder(requireActivity())

        builder
            .setView(view)
            .setPositiveButton(R.string.file_rename, this)
            .setNegativeButton(R.string.common_cancel, this)
            .setTitle(R.string.rename_dialog_title)

        return builder
    }

    private fun initAlertDialog() {
        val alertDialog = dialog as AlertDialog?

        if (alertDialog != null) {
            positiveButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE) as MaterialButton
            val negativeButton = alertDialog.getButton(AlertDialog.BUTTON_NEGATIVE) as MaterialButton

            positiveButton?.let {
                viewThemeUtils.material.colorMaterialButtonPrimaryTonal(it)
            }
            viewThemeUtils.material.colorMaterialButtonPrimaryBorderless(negativeButton)
        }
    }

    private val oCCapability: OCCapability
        get() = fileDataStorageManager.getCapability(currentAccount.user.accountName)

    override fun onClick(dialog: DialogInterface, which: Int) {
        if (which == AlertDialog.BUTTON_POSITIVE) {
            var newFileName = ""

            if (binding.userInput.text != null) {
                newFileName = binding.userInput.text.toString().trim { it <= ' ' }
            }

            val errorMessage = checkFileName(newFileName, oCCapability, requireContext(), null)
            if (errorMessage != null) {
                DisplayUtils.showSnackMessage(requireActivity(), errorMessage)
                return
            }

            if (mTargetFile?.isOfflineOperation == true) {
                fileDataStorageManager.renameCreateFolderOfflineOperation(mTargetFile, newFileName)
                if (requireActivity() is FileDisplayActivity) {
                    val activity = requireActivity() as FileDisplayActivity
                    activity.refreshFolderWithDelay()
                }
            } else {
                (requireActivity() as ComponentsGetter).fileOperationsHelper.renameFile(mTargetFile, newFileName)
            }
        }
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit

    /**
     * When user enters a hidden file name, the 'hidden file' message is shown.
     * Otherwise, the message is ensured to be hidden.
     */
    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        var newFileName = ""
        if (binding.userInput.text != null) {
            newFileName = binding.userInput.text.toString().trim { it <= ' ' }
        }

        val errorMessage = checkFileName(newFileName, oCCapability, requireContext(), fileNames)

        if (isFileHidden(newFileName)) {
            binding.userInputContainer.error = getText(R.string.hidden_file_name_warning)
        } else if (errorMessage != null) {
            binding.userInputContainer.error = errorMessage
            positiveButton?.isEnabled = false
        } else if (binding.userInputContainer.error != null) {
            binding.userInputContainer.error = null
            // Called to remove extra padding
            binding.userInputContainer.isErrorEnabled = false
            positiveButton?.isEnabled = true
        }
    }

    override fun afterTextChanged(s: Editable) = Unit

    companion object {
        private const val ARG_TARGET_FILE = "TARGET_FILE"
        private const val ARG_PARENT_FOLDER = "PARENT_FOLDER"

        /**
         * Public factory method to create new RenameFileDialogFragment instances.
         *
         * @param file File to rename.
         * @return Dialog ready to show.
         */
        @JvmStatic
        fun newInstance(file: OCFile?, parentFolder: OCFile?): RenameFileDialogFragment {
            val bundle = Bundle().apply {
                putParcelable(ARG_TARGET_FILE, file)
                putParcelable(ARG_PARENT_FOLDER, parentFolder)
            }

            return RenameFileDialogFragment().apply {
                arguments = bundle
            }
        }
    }
}
